core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt

```diff
@@ -40,26 +47,38 @@ class UnitMovementAlgorithms(val unit:MapUnit) {
         if (unit.ignoresTerrainCost) return 1f + extraCost
         if (areConnectedByRiver) return 100f  // Rivers take the entire turn to cross
 
-        if (unit.doubleMovementInForestAndJungle &&
-                (to.terrainFeatures.contains(Constants.forest) || to.terrainFeatures.contains(Constants.jungle)))
-            return 1f + extraCost // usually forest and jungle take 2 movements, so here it is 1
+        val terrainCost = to.getLastTerrain().movementCost.toFloat()
+
+        if (unit.noTerrainMovementUniques)
+            return terrainCost + extraCost
+
+        if (to.terrainFeatures.any { unit.doubleMovementInTerrain[it] == MapUnit.DoubleMovementTerrainTarget.Feature })
+            return terrainCost * 0.5f + extraCost
 
         if (unit.roughTerrainPenalty && to.isRoughTerrain())
-            return 100f // units that have to sped all movement in rough terrain, have to spend all movement in rough terrain
+            return 100f // units that have to spend all movement in rough terrain, have to spend all movement in rough terrain
         // Placement of this 'if' based on testing, see #4232
-        
+
         if (civInfo.nation.ignoreHillMovementCost && to.isHill())
             return 1f + extraCost // usually hills take 2 movements, so here it is 1
 
-        if (unit.doubleMovementInCoast && to.baseTerrain == Constants.coast)
-            return 1 / 2f + extraCost
+        if (unit.noBaseTerrainOrHillDoubleMovementUniques)
+            return terrainCost + extraCost
 
-        if (unit.doubleMovementInSnowTundraAndHills && to.isHill())
-            return 1f + extraCost // usually hills take 2
-        if (unit.doubleMovementInSnowTundraAndHills && (to.baseTerrain == Constants.snow || to.baseTerrain == Constants.tundra))
-            return 1 / 2f + extraCost
+        if (unit.doubleMovementInTerrain[to.baseTerrain] == MapUnit.DoubleMovementTerrainTarget.Base)
+            return terrainCost * 0.5f + extraCost
+        if (unit.doubleMovementInTerrain[Constants.hill] == MapUnit.DoubleMovementTerrainTarget.Hill && to.isHill())
+            return terrainCost * 0.5f + extraCost
 
-        return to.getLastTerrain().movementCost.toFloat() + extraCost // no road
+        if (unit.noFilteredDoubleMovementUniques)
+            return terrainCost + extraCost
+        if (unit.doubleMovementInTerrain.any { 
+            it.value == MapUnit.DoubleMovementTerrainTarget.Filter &&
+                to.matchesFilter(it.key)
+        })
+            return terrainCost * 0.5f + extraCost
```

Comment:
A. I think it's time we deprecated the old Hill, but that's for a different PR
B. if/else? If a unit has multiple, I don''t think they should stack. Maybe that's just me.

Reply:
They don't stack - it's peppered with `return`s. Case for a `when` but when AS doesn't whine I tend to leave the old code structure.

And yes this still uses Constants.hill outside deprecation, but that needs more thorough thinking and a clear vision how-to which I currently don't see.